working/digit-separators/feature-specification.md

```diff
@@ -0,0 +1,192 @@
+preferred formatting for print text.
+
+The apostrope is also the string single-quote character. We don't currently
+allow adjacent numbers and strings, but if we ever do, then this syntax becomes
+ambiguous. It's still possible (we disambiguate by assuming it's a digit
+separator). It is currently used by C++ 14 as a digit group separator, so it is
+definitely possible.
+
+That leaves underscore, which could be the start of an identifier. Currently
+`100_000` would be tokenized as "integer literal 100" followed by "identifier
+_000". However, users would never write an identifier adjacent to another token
+that contains identifier-valid characters (unlike strings, which have clear
+delimiters that do not occur anywher else), so this is unlikely to happen in
+practice. Underscore is already used by a large number of programming languages
+including Java, Swift, and Python.
+
+We also want to allow multiple separators for higher-level grouping, e.g.,:
+
+```none
+100__000_000_000__000_000_000
+```
+
+For this purpose, the underscore extends gracefully. So does space, but has the
+disadvantage that it collapses when inserted into HTML, whereas `''` looks odd.
+
+### Related work
+
+* [Java digit separators](https://docs.oracle.com/javase/8/docs/technotes/guides/language/underscores-literals.html)
+* [Python PEP 515 - underscores in numeric literals](https://peps.python.org/pep-0515/)
+
+### Possible new lint rules
+
+There are some possible new lint rule considerations, but none of these are
+considered vital to the usability or general success of the feature.
+
+The feature is designed to help the readability of long numbers. But a
+developer can still make an error about where to place separators. For example:
+
+```
+var one = 1_000_000;
+var two = 2_000_000;
+var three = 3_000_000;
+var four = 4_0000_000;
+```
+
+If a developer uses the Dart formatter to format their code, they cannot try to
+vertically align the numbers with whitespace (extra space characters are
+removed by the formatter). So we could offer a lint rule to only place
+separators every three digits of a decimal number. Also possibly a similar rule
+for hexadecimal numbers. If a developer ever uses digit separators for a
+different purpose (as in separating the digits of a phone number), the rule may
+not prove useful.
+
+### Possible new quick fixes
+
+There are some possible new automated fix ("quick fix") considerations, but
+none of these are considered vital to the usability or general success of the
+feature.
+
+#### Unexpected underscores
+
+With the digit-separators feature, separators can be added between _digits_ of
+a number literal, but nowhere else. In most error cases, the unexpected
+underscore can be detected as such, and we can offer quick fixes to remove
+unexpected errors (for example, `100_`, `100_e.2`, `100._00`). In a few cases,
+the intention is not as straightforward, such as `100._100`, where `_100` can
+be a legal name of an extension member.
```

Comment:
But it's also possible to check whether it's an extension (in the same library, since it's a private name), and when it most likely isn't, the fix can be recommended.

Reply:
Added